pilot/pkg/serviceregistry/kube/controller/ambient/cluster.go

```diff
@@ -0,0 +1,248 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ambient
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"time"
+
+	"go.uber.org/atomic"
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pkg/cluster"
+	"istio.io/istio/pkg/config/mesh/meshwatcher"
+	"istio.io/istio/pkg/config/schema/gvr"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/kube/kubetypes"
+	filter "istio.io/istio/pkg/kube/namespace"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/slices"
+)
+
+var _ krt.ResourceNamer = &Cluster{}
+
+type Cluster struct {
+	// ID of the cluster.
+	ID cluster.ID
+	// Client for accessing the cluster.
+	Client kube.Client
+
+	SourceSecret types.NamespacedName
+
+	kubeConfigSha [sha256.Size]byte
+	// initialSync is marked when RunAndWait completes
+	initialSync *atomic.Bool
+	// initialSyncTimeout is set when RunAndWait timed out
+	initialSyncTimeout *atomic.Bool
+	stop               chan struct{}
+	namespaces         krt.Collection[*corev1.Namespace]
+	pods               krt.Collection[*corev1.Pod]
+	services           krt.Collection[*corev1.Service]
+	endpointSlices     krt.Collection[*discovery.EndpointSlice]
+	nodes              krt.Collection[*corev1.Node]
+	gateways           krt.Collection[*v1beta1.Gateway]
+
+	// xref: https://github.com/istio/istio/pull/56097
+	// TODO: Figure out if we really want to keep doing this
+	// Note that the impl details of TestSeamlessMigration
+	// depend on this, so if we remove it, we'll need to
+	// adjust the test.
+	configmaps krt.Collection[*corev1.ConfigMap]
+
+	initialized *atomic.Bool
+}
+
+func (c *Cluster) ResourceName() string {
+	return c.ID.String()
+}
+
+func (c *Cluster) Run(mesh meshwatcher.WatcherCollection, debugger *krt.DebugHandler) {
```

Comment:
`mesh` is passed in, while others are struct members because `mesh` is local to the current istiod/config cluster, and the others are this `Cluster`. Can we add some naming or comments to clarify that?

Reply:
Renamed to `localMeshConfig`